### ISSUE TITLE
[feature request] add hook onNetworkError

### ISSUE THREAD

**Issue by xialvjun** [sindresorhus/ky#296]
just as the title
because `afterResponse` is not invoked when there is a network error.

**Comment by duongvanba** [comment]
Are there any fix ?

**Comment by szmarczak** [comment]
> Are there any fix ?

No, there is not; I don't think commenting ["is there any fix"](https://blog.sindresorhus.com/issue-bumping-e3b9740e2a0) will bring this feature forward. A PR is always welcome :)

Maybe a generic hook for all errors would be better? Like we do in [Got](https://github.com/sindresorhus/got/#hooksbeforeerror).

**Comment by duongvanba** [comment]
Thank for your reply
I'm implement de-duplicate concurrent request in my react appliction
I have a custom hook that use Ky to fetch api from server, but I use this hook many places, so I just implement a Ky hook to de-duplicate request
But i have problem when detect first request status on network error. There is no way to use Ky hooks to detect first request network error to reject a Promise for other reques
Anw, i see that Ky is awesome package
I come here from Got, I love Got and I really nodejs support HTTP2 normaly that we can use http2 as default in future project
Thank you so much for quick reply

**Comment by szmarczak** [comment]
> But i have problem when detect first request status on network error. There is no way to use Ky hooks to detect first request network error to reject a Promise for other reques

The solution is pretty simple. You need to store an array of abort controllers, and if the first request fails (you can use `promise.catch()`), call all of them.

**Comment by sholladay** [comment]
As far as I know, there hasn't been any progress at WHATWG related to discrete error types for `ferch`, which is what we would need to do this properly.

I think probably the best we can do in Ky, at least for now, is to act like any non-HTTP error is a network error. We could provide a hook for that. It would end up getting called for invalid fetch options too, not just network errors. That might be confusing. But it's doable.

**Comment by szmarczak** [comment]
> That might be confusing. But it's doable.

Not necessarily confusing, it would be pretty useful when attaching a logger.

**Comment by duongvanba** [comment]
Thank you so much

Vào Th 6, 4 thg 12, 2020 vào lúc 17:18 Szymon Marczak <
notifications@github.com> đã viết:

**Comment by MichaelSitter** [comment]
Struggled with this for the better part of a day, I'll leave this as a potential solution in case its useful. In my case I have a reusable client with circuit breaking which needs to know about timeouts & network errors, in addition to HTTP errors. Providing a wrapped fetch option at least gave me a way to observe those errors outside of hooks.

```
ky.extend({
    fetch: async (input: RequestInfo, init?: RequestInit): Promise<Response> => {
      const fn = opts.fetch || fetch
      try {
        return await fn(input, init)
      } catch(err) {
        // handle non-HttpErrors; will not trigger afterResponse hooks
        throw err
      }
    })
```

**Comment by llaz00** [comment]
I think all these issues #508 #605 share the same basis. There's a discussion needed on how to handle network errors.

**Comment by sholladay** [comment]
In the upcoming v2, this is much easier because all errors are routed through `beforeError`. See: #829

```js
import isNetworkError from 'is-network-error';
import ky from 'ky';

await ky('https://example.com', {
    hooks : {
        beforeError : [
            ({ error }) => {
                if (isNetworkError(error) {
                    console.log('Network error:', error);
                }
            }
        ]
    }
});
```

I'm going to close this since I think it's solved in spirit. Let me know if you feel strongly that there should be a dedicated hook and we can discuss it further.
